torch/distributed/_spmd/graph_optimization.py

```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+            # is not empty. If the original argument list is empty, the new
+            # one must be an empty list as well.
+            if orig_arg:
+                optim_arg.append(orig_arg[idx])
+
+        # If argument order of step is the same as optimizer, nothing has to be
+        # done. However, it is risky to rely on this assumption so we populate
+        # the orig_step_indices.
+        old_step_output = optim_args[group_idx].state_steps[-1]
+        assert str(old_step_output.target).startswith(
+            "aten.copy_"
+        ), f"The copy output is {old_step_output.target}, expect aten.copy_"
+        old_step_getitem = old_step_output.args[1]
+        assert "getitem" in str(
+            old_step_getitem.target
+        ), f"The copy getitem is {old_step_getitem.target}, expect operator.getitem"
+        old_step_idx = old_step_getitem.args[1]
+        orig_step_indices[group_idx].append(old_step_idx)
+
+    if not all(l for l in (orig_step_indices + orig_optim_indices)):
+        raise ValueError("At least one split optimizer does not have input.")
+
+    output = get_output(gm.graph)
+    result: List[FusedOptimizerBlock] = []
+    flatten_output_args, spec = tree_flatten((output.args, output.kwargs))
+    flatten_output_args_indices: DefaultDict[
+        fx.Node, Set[int]
+    ] = collections.defaultdict(set)
+    for idx, output_arg in enumerate(flatten_output_args):
+        if isinstance(output_arg, fx.Node):
+            flatten_output_args_indices[output_arg].add(idx)
+
+    def replace_flatten_output_args(orig_node: fx.Node, new_node: fx.Node):
+        for idx in flatten_output_args_indices[orig_node]:
+            flatten_output_args[idx] = new_node
+
+    # Create the new step and optim nodes and blocks.
+    for group_idx in range(2):
+        step_args: List[fx.Node] = []
+        orig_step_outputs: List[fx.Node] = []
+        # We have to create the new step node and block first because it is used
+        # for the new optim node as the input.
+        with gm.graph.inserting_after(orig_optim_block.optim.optim_node):
+            for idx in orig_step_indices[group_idx]:
+                step_args.append(
+                    cast(Tuple[fx.Node, ...], orig_optim_block.step.add_node.args[0])[
+                        idx
+                    ]
+                )
+                orig_step_outputs.append(orig_optim_block.step.outputs[idx])
+            step = gm.graph.call_function(
+                torch.ops.aten._foreach_add.Scalar,
+                (step_args, 1),
+            )
+        step_block = ForeachAddBlock(step, generate_output=True)
+        for i, step_output in enumerate(step_block.outputs):
+            # Replace the original step output in the graph output node with
+            # the new one.
+            orig_step_output = orig_step_outputs[i]
+            replace_flatten_output_args(orig_step_output, step_output)
```

Comment:
do we need to worry about there might be other nodes other then output consuming step out?

Reply:
I was thinking to use `fx.Node.replace_all_uses_with()`. However, this will perform 6*O(grad) tree_unflatten/tree_flatten. So I decided to use write `replace_flatten_output_args` to speed up the performance and assume only output as the user. But you are right about the case. 

Let me use `fx.Node.replace_all_uses_with()` and add a TODO to improve the performance.